streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreChangelogReaderTest.java

```diff
@@ -44,495 +44,733 @@
 
-        expect(active.restoringTaskFor(topicPartition)).andReturn(task);
-        replay(active);
+        changelogReader.transitToUpdateStandby();
 
-        changelogReader.restore();
-        assertThat(callback.restored.size(), equalTo(5));
-    }
+        assertEquals(StoreChangelogReader.ChangelogState.RESTORING, changelogReader.changelogMetadata(tp).state());
+        assertEquals(StoreChangelogReader.ChangelogState.RESTORING, changelogReader.changelogMetadata(tp1).state());
+        assertEquals(StoreChangelogReader.ChangelogState.RESTORING, changelogReader.changelogMetadata(tp2).state());
+        assertEquals(mkSet(tp, tp1, tp2), consumer.assignment());
+        assertEquals(Collections.emptySet(), consumer.paused());
 
-    @Test
-    public void shouldNotThrowTaskMigratedExceptionIfEndOffsetNotExceededDuringRestoreForSourceTopicEOSEnabled() {
-        final int totalMessages = 10;
-        setupConsumer(totalMessages, topicPartition);
-        // EOS enabled simulated by setting offsets with commit marker
-        // records have offsets 0..9 10 is commit marker so 11 is ending offset
-        consumer.updateEndOffsets(Collections.singletonMap(topicPartition, 11L));
+        // transition to update standby is NOT idempotent
+        assertThrows(IllegalStateException.class, changelogReader::transitToUpdateStandby);
 
-        changelogReader.register(topicPartition, stateManager);
+        changelogReader.remove(Collections.singletonList(tp));
+        changelogReader.register(tp, activeStateManager);
 
-        expect(active.restoringTaskFor(topicPartition)).andReturn(task);
-        replay(active);
+        // if a new active is registered, we should immediately transit to standby updating
+        assertThrows(IllegalStateException.class, changelogReader::restore);
 
-        changelogReader.restore();
-        assertThat(callback.restored.size(), equalTo(10));
+        assertEquals(StoreChangelogReader.ChangelogState.RESTORING, changelogReader.changelogMetadata(tp).state());
+        assertEquals(StoreChangelogReader.ChangelogState.RESTORING, changelogReader.changelogMetadata(tp1).state());
+        assertEquals(StoreChangelogReader.ChangelogState.RESTORING, changelogReader.changelogMetadata(tp2).state());
+        assertEquals(mkSet(tp, tp1, tp2), consumer.assignment());
+        assertEquals(Collections.emptySet(), consumer.paused());
+
+        changelogReader.transitToRestoreActive();
+        assertEquals(mkSet(tp, tp1, tp2), consumer.assignment());
+        assertEquals(mkSet(tp1, tp2), consumer.paused());
     }
 
     @Test
-    public void shouldRestoreUpToOffsetLimit() {
-        setupConsumer(10, topicPartition);
-        changelogReader.register(topicPartition, stateManager);
-        expect(active.restoringTaskFor(topicPartition)).andStubReturn(task);
-        replay(active, task);
-        changelogReader.restore();
+    public void shouldThrowIfRestoreCallbackThrows() {
```

Comment:
nit: for any task type neutral tests, I would feel more comfortable if we could allow testing for both active and standby state managers in two separate tests, or share the same skeleton somehow

Reply:
I've made the test parameterized.